docs(az.service-bus): restructure az service-bus messaging (#546)
arcus-azure/arcus.messaging: docs/preview/03-Features/01-Azure/01-service-bus.md

        {
            // Adds a filter to the message handler registration:
            // Only messages with 'Type=Order' property goes through this message handler.
            options.AddMessageContextFilter(ctx => ctx.Properties["Type"] == "Order");

            // Adds a filter to the message handler registration:
            // Only messages with certain bodies goes through this message handler.
            options.AddMessageBodyFilter((Order order) => order.OrderId = "123");

            // Adds a custom message deserializer to the message handler registration:
            // Only messages that gets deserialized successfully goes through this message handler.
            // 👉 Custom implementations of the `Arcus.Messaging.Abstractions.MessageHandling.IMessageBodySerializer` interface.
            //      public interface IMessageBodySerializer
            //      {
            //          Task<MessageResult> DeserializeMessageAsync(string messageBody);
            //      }
            options.AddMessageBodySerializer(new CustomXmlMessageBodySerializer());
        });
```

### Custom message settlement

> When messages can't be matched to any of your custom registered message handlers, Arcus will dead-letter the message. When one of your message handlers fails to process a message, it will get abandoned and maybe only after a third try be dead-lettered.

Reviewer: ```suggestion
When messages can't be matched to any of your custom registered message handlers, Arcus will dead-letter the message. When one of your message handlers fails to process a message, it will get abandoned and only after the maximum number of redeliveries has been reached, the message will be dead-lettered.
```

> This settlement of received Azure Service Bus messages can also be customized by calling one of the Service Bus operations yourself via the message context.

Reviewer: Maybe we should mention here that it would be good practice for the application developer to deadletter the message himself when a non-transient/non retryable error occurs ?

Author: ```suggestion
This settlement of received Azure Service Bus messages can also be customized by calling one of the Service Bus operations yourself via the message context.

> 🎖️ It is a good practice as application developer to dead-letter the message yourself when a non-transient/fatal error occurs.
```

Author: Done in https://github.com/arcus-azure/arcus.messaging/pull/546/commits/4da16920a98830b40b01cfac82f32c6ca7c38159

```csharp
public class OrderMessageHandler : IAzureServiceBusMessageHandler<Order>
{
    public async Task Task ProcessMessageAsync(
        Order message,
        AzureServiceBusMessageContext messageContext,
        MessageCorrelationInfo correlation,
        CancellationToken cancellation)
    {
        await messageContext.DeadLetterMessageAsync("Reason: Unsupported", "Message type is not supported");
    }
}
```

The following operations are supported:
* **Dead-letter**
* **Abandon**

> * **Complete**

Author: Also added the custom message settlement via the message context, added in PR #551

﻿---
sidebar_label: Service bus
---

# Azure Service Bus messaging

> The `Arcus.Messaging.Pumps.ServiceBus` library provides a way to process Azure Service bus messages on queues/topic subscriptions via custom routed *message handlers*, instead of interacting with the [`ServiceBusReceiver`](https://learn.microsoft.com/en-us/dotnet/api/azure.messaging.servicebus.servicebusreceiver) yourself. 

Reviewer: ```suggestion
The `Arcus.Messaging.Pumps.ServiceBus` library provides a way to process Azure Service Bus messages on queues/topic subscriptions via custom routed *message handlers*, instead of interacting with the [`ServiceBusReceiver`](https://learn.microsoft.com/en-us/dotnet/api/azure.messaging.servicebus.servicebusreceiver) yourself. 
```

Microsoft's documentation always mentions 'Azure Service **B**us', so I'd do that here as well.

            });
    });
```

### Register your message handlers on message pump
Your custom *message handlers* can be registered on a *message pump* registration. Both the type of the handler and type of of messages it can process, is required.

```csharp
using Microsoft.Extensions.DependencyInjection;

Host.CreateDefaultBuilder()
    .ConfigureServices(services =>
    {
        services.AddServiceBusTopicMessagePump(...)
                .WithServiceBusMessageHandler<OrderMessageHandler, Order>()
                .WithServiceBusMessageHandler<OrderV2MessageHandler, OrderV2>();
    });
```

> **⚠️ Considerations:**
> * The order in which the *message handlers* are registered matters when a message is routed.
> * Only the first matching *message handler* will be used to route the message, even when multiple match.
> * Multiple *message handlers* with the same (message) type can registered, but they need to distinguish themselves with [routing options](#message-handler-routing-customization).

## Customization
Due to the wide range of situations within messaging solutions, Arcus Messaging supports highly customizable message pump/handler registrations.

> ### Message pump customization

Reviewer: customization or rather configuration ?

Author: In a way, all options here is to 'customize' additional functionality, no?

Reviewer: In my head, customization goes a bit further than configuration, but I'm ok with it.

            // Adds a custom message deserializer to the message handler registration:
            // Only messages that gets deserialized successfully goes through this message handler.
            // 👉 Custom implementations of the `Arcus.Messaging.Abstractions.MessageHandling.IMessageBodySerializer` interface.
            //      public interface IMessageBodySerializer
            //      {
            //          Task<MessageResult> DeserializeMessageAsync(string messageBody);
            //      }
            options.AddMessageBodySerializer(new CustomXmlMessageBodySerializer());
        });
```

### Message handler fallback customization
When a received Azure Service bus message cannot be routed to any of the registered *message handlers*, then the message will be [dead-lettered](https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-dead-letter-queues) automatically.

Arcus Messaging provides a 'hook' when such a scenario occurs, called: 'fallback message handlers'. These special kind of *message handlers* are only used in scenarios where the 'general' handlers can't be used.

> 

Author: Yes, let's remove that indeed. 👍

> 💡 This is a great way to do custom dead-lettering.

> To implement such a 'fallback' *message handler*, you can use the `AzureServiceBusFallbackMessageHandler` abstract class. This gives you access to `DeadLetterMessageAsync` and other Service bus-specific methods.

Author: Yep, indeed. We can remove this, but indeed, would be more comfortable in another PR (maybe even in the #551 once this is merged).
